### PR TITLE
Remove bindgen dependency

### LINKED ISSUE #31: Removing build-time dependency on bindgen
Hello!  We're interested in using `libffi` from Rust but would ideally like to avoid the bulid-time dependency on `bindgen` and libclang.

Other `*-sys` crates seem to not depend on bindgen, for example [`llvm-sys`](https://crates.io/crates/llvm-sys).  Is there a reason why this is necessary in the libffi case?  If not, would you all be interested in having this feature in `libffi-sys-rs`?

### PR DESCRIPTION
**Pull request by yorickpeterse**
This removes the dependency on bindgen, replacing it with a set of bindings directly included in the repository. The bindings were generated on Linux, cleaned up, then compared with the bindgen output for MacOS and Windows (using MSYS2). Apart from some noise caused by unused constants, the libffi API code was the same.

Removing bindgen cuts down the amount of dependencies, most notably the dependency on libclang. Removing this simplifies the installation procedure, as well as simplifying the build scripts a bit.

The complex FFI types are hidden behind a "complex" feature flag that is disabled by default, similar to the libffi-rs crate. The longdouble type is disabled on GNU/Linux ARM devices, per https://github.com/tov/libffi-sys-rs/pull/37#issuecomment-701645235 and https://github.com/tov/libffi-rs/pull/14.

This fixes https://github.com/tov/libffi-sys-rs/issues/31.

## Testing results

| Platform  | Arch   | `cargo test` | `cargo test --features system`
|:----------|:-------|:-------------|:------------------------------
| Linux     | x86-64 | :ok:         | :ok:
| macOS     | x86-64 | :ok:         | :ok:
| MSYS2     | x86-64 | :ok:         | :warning: Unable to find the library, same on `master`
| MSVC      | x86-64 | :ok:         | :ok:
| Linux     | armv7  | :ok:         | :ok:

### CODE REVIEW

File: build/msvc.rs [hunk 1]
```diff
@@ -51,27 +44,28 @@ pub fn build_and_link() -> IncludePaths {
         .warnings(false)
         .compile("libffi");
 
-    IncludePaths(INCLUDE_DIRS.iter().map(Into::into).collect())
+    //IncludePaths(INCLUDE_DIRS.iter().map(Into::into).collect())
+}
+
+pub fn probe_and_link() {
+    // TODO
```
Comment (yorickpeterse) [inline build/msvc.rs]: @timfish Does MSVC support dynamic linking of libffi? If so, how do we implement that here as I'm not sure the pkg-config crate works on Windows. If not, then I think we can just make this function an alias of `build_and_link`.

File: build/msvc.rs [hunk 2]
```diff
@@ -1,11 +1,6 @@
 use crate::common::*;
```
Comment (yorickpeterse) [inline build/msvc.rs]: TODO: undo this formatting caused by rustfmt, to reduce noise.

File: src/binding.rs [hunk 3]
```diff
@@ -0,0 +1,350 @@
+/// Raw bindings for the libffi C API.
```
Comment (yorickpeterse) [inline src/binding.rs]: TODO: add some actual libffi tests (e.g. that just test a math function)

File: build/msvc.rs (hunk 1 above)
Comment (timfish) [reply build/msvc.rs]: Sorry no idea, I've never tried!

**Comment by timfish** [comment]
It's worth pointing out that the bindings are slightly different on ARM because there is no `longdouble`.

Checkout this PR to try and fix it: https://github.com/tov/libffi-rs/pull/14

But other than that it seems like a great idea. I've been dealing with version issues lately caused by conflicted versions of libclang.

File: build/msvc.rs (hunk 1 above)
Comment (yorickpeterse) [reply build/msvc.rs]: @timfish No worries. I'll change the code to just call `build_and_link()`, that way we don't break any MSVC setups. We can always change this in the future if necessary.

**Comment by yorickpeterse** [comment]
> It's worth pointing out that the bindings are slightly different on ARM because there is no `longdouble`.
> 
> Checkout this PR to try and fix it: [tov/libffi-rs#14](https://github.com/tov/libffi-rs/pull/14)
> 
> But other than that it seems like a great idea. I've been dealing with version issues lately caused by conflicted versions of libclang.

@timfish I don't have any ARM hardware to test this on, but by the looks of it we can just copy the `cfg` attributes to this PR. Do you happen to have any ARM devices you could test those changes on?

**Comment by yorickpeterse** [comment]
@timfish I have adopted the `cfg` checks from your PR, and added a `complex` feature flag for the complex FFI types. If you have the opportunity, please give this PR a try on ARM and MSVC (if you have access to such a setup).

**Comment by timfish** [comment]
Yep, I've got a Windows machine and a Raspberry Pi that I'll give this a try on.

**Comment by yorickpeterse** [comment]
Testing results thus far:

| Platform  | Arch   | `cargo test` | `cargo test --features system`
|:----------|:-------|:-------------|:------------------------------
| Linux     | x86-64 | :ok:         | :ok:
| macOS     | x86-64 | :ok:         | :ok:
| MSYS2     | x86-64 | :ok:         | :warning: 

Testing was done by running `cargo test` and `cargo test --features system`. Using the system feature on MSYS2 produces linker errors (seems the .dll can't be found), which I'm currently looking into.

**EDIT 1:** This is also happening when you run `cargo test --features system` on `master`, so either this is an existing issue or my MSYS2 setup is missing something.

**EDIT 2:** It seems that on MSYS2 the DLL is called `libffi-7.dll`, and there's no `libffi.dll` to be found.

**Comment by timfish** [comment]
| Platform | Arch | `cargo test` | `cargo test --features system`
|--------|----------|-----|------|
| MSVC |	x86-64 |	🆗|	🆗<sup>1</sup> |
| Linux |	armv7 |	🆗|	🆗 |

1. I think this is a false positive (and it's still doing static linking?) because I don't have libffi binaries anywhere on that machine.

**Comment by yorickpeterse** [comment]
@timfish 

> I think this is a false positive (and it's still doing static linking?) because I don't have libffi binaries anywhere on that machine.

Correct, it static links the source install of libffi. I have no idea how dynamic linking and library management works under MSVC, or how you'd even install libffi.

**Comment by timfish** [comment]
I think on Windows you can use [Vcpkg](https://github.com/microsoft/vcpkg) but I wouldn't worry about it.  I can't see anybody using it when there's reliable static linking.

**Comment by yorickpeterse** [comment]
I've been unable to get dynamic linking on MSYS2 to work, even when using the Rust provided by `mingw-w64-x86_64-rust`. Since this is already happening on `master` I think this may be an MSYS2 issue, or perhaps Rust isn't able to find the right libraries (even when I explicitly specify them).

Since this is not a new issue, I think we can leave this as-is. As far as I understand, static linking on Windows is preferred anyway.

**Comment by yorickpeterse** [comment]
@tov could you take a look at these changes? Thanks!

p.s. If you need help maintaining the crates and such, please let me know; I'm more than happy to help increase the bus factor :smiley:

**Comment by yorickpeterse** [comment]
For those that want to try this pull request by patching their `libffi-sys` version, you can do so by adding the following to your `Cargo.toml`:

```toml
[patch.crates-io]
libffi-sys = { git = "https://github.com/YorickPeterse/libffi-sys-rs.git", branch = "no-bindgen-patch" }
libffi = { git = "https://github.com/YorickPeterse/libffi-rs.git", branch = "no-bindgen-patch" }
```

These branches contain these changes in this PR, in addition to a version change so they can be used to patch existing installations. To apply them, you need to run `cargo update -p libffi-sys` and `cargo update -p libffi`, and you should be good to go.

**Comment by yorickpeterse** [comment]
Worth mentioning: with these changes I can reduce my clean `cargo build --release` times by about 1 minute. That's quite significant, and far more than what I expected. Note that this is using the `system` feature, so it doesn't include the time it takes to compile libffi itself. This was also tested on a laptop, so the timings are not the most reliable.

**Comment by yorickpeterse** [comment]
@tov Do you have time to look at this PR any time soon? Thanks!

**Comment by tov** [comment]
Hey! Sorry. I am going to make you a maintainer, @YorickPeterse. I very much appreciate the offer.

**Comment by tov** [comment]
Inlining all the magic numbers from `libffi.h` frightens me! But you are using this library for real and I’m not, so I think it makes sense for me to defer to your judgment here. And not depending on bindgen is excellent.

**Comment by yorickpeterse** [comment]
@tov Thanks for adding me! Unfortunately there's no other way to provide these constants, as they are all created using `#define` in `ffi.h`, instead of being globals. I don't think these will ever have their value changed though, so it should be fine.

**Comment by yorickpeterse** [comment]
Since there don't seem to be any particular objections against these changes, I believe we can go ahead and merge them.

**Pull request merged**